File: android/walletlink/src/main/java/com/coinbase/walletlink/dtos/Web3ResponseDTO.kt
package com.coinbase.walletlink.dtos

import com.coinbase.wallet.core.util.JSON
import com.coinbase.walletlink.models.RequestMethod
import com.squareup.moshi.Types

internal class Web3ResponseDTO<T> private constructor(
    val id: String,
    val response: Web3Response<T>
) {
    constructor(
        id: String,
        method: RequestMethod,
        result: T
    ) : this(
        id = id,
        response = Web3Response(method, result, null)
    )

    constructor(
        id: String,
        method: RequestMethod,
        errorMessage: String?
    ) : this(
        id = id,
        response = Web3Response(method, null, errorMessage)
    )
}

internal inline fun <reified T> Web3ResponseDTO<T>.asJsonString(): String {
    val type = Types.newParameterizedType(Web3ResponseDTO::class.java, T::class.java)
    val adapter = JSON.moshi.adapter<Web3ResponseDTO<T>>(type)

    return adapter.toJson(this)
}

internal data class Web3Response<T>(val method: RequestMethod, val result: T?, val errorMessage: String?)

File: ios/Source/DTOs/Web3ResponseDTO.swift
// Copyright (c) 2017-2019 Coinbase Inc. See LICENSE

import Foundation

struct Web3ResponseDTO<T: Codable>: Codable, JSONSerializable {
    let type: String = "WEB3_RESPONSE"
    let id: String
    let response: Web3Response<T>

    init(id: String, method: RequestMethod, result: T) {
        self.id = id
        response = Web3Response<T>(method: method, result: result, errorMessage: nil)
    }

    init(id: String, method: RequestMethod, errorMessage: String) {
        self.id = id
        response = Web3Response<T>(method: method, result: nil, errorMessage: errorMessage)
    }
}

struct Web3Response<T: Codable>: Codable {
    let method: RequestMethod
    let result: T?
    let errorMessage: String?
}
